File: backend/app/schemas/salary_structure.py
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from pydantic import BaseModel, Field, field_validator
from app.schemas.salary_rule import SalaryRuleBase, SalaryRuleResponse, SalaryRuleCreate


class SalaryStructureBase(BaseModel):
    name: str = Field(..., min_length=2, max_length=100, description="Salary structure name e.g. Standard Monthly Salary")
    code: str = Field(..., min_length=1, max_length=50, description="Unique structure code e.g. STANDARD_MONTHLY")
    company: str = Field("PeoplePay360 Inc.", max_length=100, description="Company name")
    pay_frequency: str = Field("monthly", description="Pay frequency: monthly, weekly, bi-weekly, semi-monthly")
    description: Optional[str] = Field(None, description="Description or applicability notes")
    effective_from: Optional[date] = Field(None, description="Effective start date")
    effective_to: Optional[date] = Field(None, description="Effective end date")
    active: bool = Field(True, description="Active status")

    @field_validator("code")
    @classmethod
    def validate_code(cls, v: str) -> str:
        clean = v.strip().upper()
        if not clean:
            raise ValueError("Structure code cannot be empty.")
        if not clean.replace("_", "").isalnum():
            raise ValueError("Structure code must contain only alphanumeric characters and underscores.")
        return clean

    @field_validator("pay_frequency")
    @classmethod
    def validate_frequency(cls, v: str) -> str:
        valid = {"monthly", "weekly", "bi-weekly", "semi-monthly"}
        cleaned = v.strip().lower()
        if cleaned not in valid:
            raise ValueError(f"Pay frequency '{v}' is invalid. Supported options: {', '.join(valid)}")
        return cleaned

    @field_validator("effective_to")
    @classmethod
    def validate_dates(cls, effective_to: Optional[date], info) -> Optional[date]:
        effective_from = info.data.get("effective_from")
        if effective_from and effective_to and effective_to < effective_from:
            raise ValueError("Effective To date cannot be earlier than Effective From date.")
        return effective_to


class SalaryStructureCreate(SalaryStructureBase):
    rules: Optional[List[SalaryRuleCreate]] = Field(None, description="Optional list of initial salary rules to create with structure")


class SalaryStructureUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    company: Optional[str] = None
    pay_frequency: Optional[str] = None
    description: Optional[str] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    active: Optional[bool] = None

    @field_validator("code")
    @classmethod
    def validate_code(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            clean = v.strip().upper()
            if not clean.replace("_", "").isalnum():
                raise ValueError("Structure code must contain only alphanumeric characters and underscores.")
            return clean
        return v

    @field_validator("pay_frequency")
    @classmethod
    def validate_frequency(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            valid = {"monthly", "weekly", "bi-weekly", "semi-monthly"}
            cleaned = v.strip().lower()
            if cleaned not in valid:
                raise ValueError(f"Pay frequency '{v}' is invalid. Supported options: {', '.join(valid)}")
            return cleaned
        return v


class SalaryStructureStatusUpdate(BaseModel):
    active: bool = Field(..., description="Target active status")


class SalaryStructureResponse(BaseModel):
    id: int
    name: str
    code: str
    company: str
    pay_frequency: str
    description: Optional[str] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    active: bool
    rules_count: int = 0
    contracts_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class SalaryStructureDetailResponse(SalaryStructureResponse):
    rules: List[SalaryRuleResponse] = []


class SalaryStructureListResponse(BaseModel):
    items: List[SalaryStructureResponse]
    total: int
    page: int
    limit: int


# ==========================================
# Salary Preview & Live Engine Models
# ==========================================
class SalaryPreviewRequest(BaseModel):
    contract_wage: float = Field(..., ge=0, description="Monthly base contract wage (e.g. 50000)")
    days_in_period: int = Field(30, ge=1, le=31, description="Total days in the payroll period")
    worked_days: float = Field(30.0, ge=0, le=31, description="Number of days worked")
    paid_leave_days: float = Field(0.0, ge=0, le=31, description="Approved paid leave days")
    unpaid_leave_days: float = Field(0.0, ge=0, le=31, description="Approved unpaid leave days")
    employee_id: Optional[int] = Field(None, description="Optional employee ID to pull real attendance/time off context")


class SalaryComponentPreview(BaseModel):
    sequence: int
    rule_id: Optional[int] = None
    rule_code: str
    rule_name: str
    category: str  # Basic, Allowance, Gross, Deduction, Employer Contribution, Net
    computation_type: str
    amount: float
    description: Optional[str] = None
    appears_on_payslip: bool = True
    employer_cost_flag: bool = False
    condition_applied: bool = True


class SalaryPreviewResponse(BaseModel):
    structure_id: Optional[int] = None
    structure_name: Optional[str] = None
    structure_code: Optional[str] = None
    contract_wage: float
    basic_salary: float
    gross_earnings: float
    total_deductions: float
    net_salary: float
    employer_cost: float
    components: List[SalaryComponentPreview]
    warnings: List[str] = []
    context_used: Dict[str, Any] = {}


class LiveComputationRuleInput(SalaryRuleBase):
    id: Optional[int] = None


class LiveComputationRequest(BaseModel):
    contract_wage: float = Field(..., ge=0)
    days_in_period: int = Field(30, ge=1, le=31)
    worked_days: float = Field(30.0, ge=0, le=31)
    paid_leave_days: float = Field(0.0, ge=0, le=31)
    unpaid_leave_days: float = Field(0.0, ge=0, le=31)
    rules: List[LiveComputationRuleInput] = []

File: frontend/src/types/salaryStructure.ts
export type RuleCategory = 
  | 'Basic'
  | 'Allowance'
  | 'Gross'
  | 'Deduction'
  | 'Employer Contribution'
  | 'Net';

export type ComputationType = 'fixed' | 'percentage' | 'formula';

export type ConditionType = 'always' | 'conditional';

export interface SalaryRule {
  id: number;
  structure_id: number;
  structure_name?: string;
  structure_code?: string;
  name: string;
  code: string;
  category: RuleCategory;
  sequence: number;
  computation_type: ComputationType;
  fixed_amount: number;
  percentage_base_code?: string | null;
  percentage_rate: number;
  formula_expression?: string | null;
  condition_type: ConditionType;
  condition_formula?: string | null;
  appears_on_payslip: boolean;
  employer_cost_flag: boolean;
  active: boolean;
  description?: string | null;
  created_at?: string;
  updated_at?: string;
}

export interface SalaryStructure {
  id: number;
  name: string;
  code: string;
  company: string;
  pay_frequency: 'monthly' | 'weekly' | 'bi-weekly' | 'semi-monthly' | string;
  description?: string | null;
  effective_from?: string | null;
  effective_to?: string | null;
  active: boolean;
  rules_count?: number;
  contracts_count?: number;
  rules?: SalaryRule[];
  created_at?: string;
  updated_at?: string;
}

export interface SalaryStructureCreatePayload {
  name: string;
  code: string;
  company?: string;
  pay_frequency?: string;
  description?: string;
  effective_from?: string;
  effective_to?: string;
  active?: boolean;
  rules?: Partial<SalaryRule>[];
}

export interface SalaryStructureUpdatePayload {
  name?: string;
  code?: string;
  company?: string;
  pay_frequency?: string;
  description?: string;
  effective_from?: string;
  effective_to?: string;
  active?: boolean;
}

export interface SalaryRuleCreatePayload {
  structure_id: number;
  name: string;
  code: string;
  category: RuleCategory;
  sequence: number;
  computation_type: ComputationType;
  fixed_amount?: number;
  percentage_base_code?: string;
  percentage_rate?: number;
  formula_expression?: string;
  condition_type?: ConditionType;
  condition_formula?: string;
  appears_on_payslip?: boolean;
  employer_cost_flag?: boolean;
  active?: boolean;
  description?: string;
}

export interface SalaryRuleUpdatePayload {
  name?: string;
  code?: string;
  category?: RuleCategory;
  sequence?: number;
  computation_type?: ComputationType;
  fixed_amount?: number;
  percentage_base_code?: string;
  percentage_rate?: number;
  formula_expression?: string;
  condition_type?: ConditionType;
  condition_formula?: string;
  appears_on_payslip?: boolean;
  employer_cost_flag?: boolean;
  active?: boolean;
  description?: string;
}

export interface SalaryComponentPreview {
  sequence: number;
  rule_id?: number;
  rule_code: string;
  rule_name: string;
  category: RuleCategory;
  computation_type: ComputationType;
  amount: number;
  description?: string | null;
  appears_on_payslip: boolean;
  employer_cost_flag: boolean;
  condition_applied: boolean;
}

export interface SalaryPreviewRequest {
  contract_wage: number;
  days_in_period?: number;
  worked_days?: number;
  paid_leave_days?: number;
  unpaid_leave_days?: number;
  employee_id?: number;
}

export interface SalaryPreviewResponse {
  structure_id?: number;
  structure_name?: string;
  structure_code?: string;
  contract_wage: number;
  basic_salary: number;
  gross_earnings: number;
  total_deductions: number;
  net_salary: number;
  employer_cost: number;
  components: SalaryComponentPreview[];
  warnings: string[];
  context_used: Record<string, any>;
}

export interface LiveComputationRequest {
  contract_wage: number;
  days_in_period?: number;
  worked_days?: number;
  paid_leave_days?: number;
  unpaid_leave_days?: number;
  rules: Partial<SalaryRule>[];
}

export interface SalaryStructureListResponse {
  items: SalaryStructure[];
  total: number;
  page: number;
  limit: number;
}

export interface SalaryRuleListResponse {
  items: SalaryRule[];
  total: number;
  page: number;
  limit: number;
}

// ==========================================
// Salary Rules Engine Calculation Preview Types
// ==========================================
export interface PayrollCalculatePreviewRequest {
  employee_id: number;
  period_start: string;
  period_end: string;
  contract_id?: number;
  custom_inputs?: Record<string, number>;
}

export interface RuleExecutionTraceItem {
  sequence: number;
  rule_id?: number;
  rule_code: string;
  rule_name: string;
  category: string;
  computation_type: string;
  status: 'calculated' | 'skipped' | 'error';
  condition_applied: boolean;
  inputs_used: Record<string, any>;
  formula_or_rate?: string | null;
  amount: number;
  appears_on_payslip: boolean;
  employer_cost_flag: boolean;
  note?: string | null;
}

export interface SalaryComponentResult {
  sequence: number;
  rule_id?: number;
  rule_code: string;
  rule_name: string;
  category: string;
  computation_type: string;
  amount: number;
  description?: string | null;
  appears_on_payslip: boolean;
  employer_cost_flag: boolean;
  condition_applied: boolean;
}

export interface EmployeeSummary {
  id: number;
  name: string;
  employee_code: string;
  work_email?: string | null;
  department: string;
  job_position: string;
  company?: string | null;
}

export interface ContractSummary {
  id: number;
  contract_code: string;
  name: string;
  wage_per_month: number;
  start_date: string;
  end_date?: string | null;
  status: string;
  salary_structure_id: number;
}

export interface StructureSummary {
  id: number;
  name: string;
  code: string;
  pay_frequency: string;
  active: boolean;
}

export interface AttendanceSummary {
  scheduled_days: number;
  worked_days: number;
  absent_days: number;
  total_worked_hours: number;
  overtime_hours: number;
  late_minutes: number;
}

export interface TimeOffSummary {
  paid_leave_days: number;
  unpaid_leave_days: number;
  total_leave_days: number;
  leave_records_count: number;
}

export interface PeriodSummary {
  start_date: string;
  end_date: string;
  days_in_period: number;
}

export interface PayrollCalculatePreviewResponse {
  mode: string;
  status: 'SUCCESS' | 'WARNING' | 'FAILED' | string;
  employee: EmployeeSummary;
  contract: ContractSummary;
  structure: StructureSummary;
  period: PeriodSummary;
  attendance: AttendanceSummary;
  time_off: TimeOffSummary;
  basic_salary: number;
  allowance_total: number;
  earning_total: number;
  gross_salary: number;
  total_deductions: number;
  net_salary: number;
  employer_contribution_total: number;
  employer_cost: number;
  components: SalaryComponentResult[];
  calculation_trace: RuleExecutionTraceItem[];
  warnings: string[];
  context_used: Record<string, any>;
}

